File: hermes-proxy/proxy.py
#!/usr/bin/env python3
"""
Hermes Multi-Model Proxy
A Python HTTP proxy server that manages multiple AI models for Hermes Agent.
Listens on port 18793 by default.
"""

import os
import sys
import json
import time
import logging
from datetime import datetime
from pathlib import Path
from functools import wraps

import yaml
import requests
from flask import Flask, request, jsonify, Response, stream_with_context

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
logger = logging.getLogger(__name__)

# Initialize Flask app
app = Flask(__name__)
app.config['JSON_AS_ASCII'] = False

# Configuration
PORT = int(os.environ.get('PORT', 18793))
HOME = os.environ.get('HOME', str(Path.home()))
CONFIG_YAML_PATH = os.path.join(HOME, '.hermes', 'config.yaml')
ROUTING_MODE_FILE = os.path.join(HOME, '.hermes-proxy', 'routing-mode.json')

# In-memory switch history
switch_history = []
MAX_HISTORY = 50

# Routing mode
routing_mode = 'codex'  # 'codex' | 'config' | 'both'


def load_routing_mode():
    """Load routing mode from file."""
    global routing_mode
    try:
        if os.path.exists(ROUTING_MODE_FILE):
            with open(ROUTING_MODE_FILE, 'r') as f:
                data = json.load(f)
                if data.get('mode') in ['codex', 'config', 'both']:
                    routing_mode = data['mode']
    except Exception as e:
        logger.warning(f"Failed to load routing mode: {e}")
        routing_mode = 'codex'


def update_routing_mode(mode):
    """Update routing mode to file using atomic rename to prevent race conditions."""
    global routing_mode
    try:
        os.makedirs(os.path.dirname(ROUTING_MODE_FILE), exist_ok=True)
        tmp_file = ROUTING_MODE_FILE + '.tmp'
        with open(tmp_file, 'w') as f:
            json.dump({'mode': mode}, f, indent=2)
        os.replace(tmp_file, ROUTING_MODE_FILE)
        routing_mode = mode
        logger.info(f"Routing mode updated to: {mode}")
        return True
    except Exception as e:
        # Clean up temp file on failure
        try:
            os.unlink(ROUTING_MODE_FILE + '.tmp')
        except OSError:
            pass
        logger.error(f"Failed to update routing mode: {e}")
        return False


def find_config_yaml():
    """Find the config.yaml file."""
    candidates = [
        os.path.join(HOME, '.hermes', 'config.yaml'),
        os.path.join(HOME, 'Library', 'Containers', 'app.nousresearch.hermes', 'Data', '.hermes', 'config.yaml'),
        os.path.join(HOME, '.hermes', 'config.yml'),
    ]

    for candidate in candidates:
        if os.path.exists(candidate):
            logger.info(f"Found config file: {candidate}")
            return candidate

    return candidates[0]  # Return first candidate as default


def parse_config_yaml(file_path):
    """Parse YAML config file."""
    try:
        with open(file_path, 'r') as f:
            return yaml.safe_load(f)
    except Exception as e:
        logger.error(f"Failed to parse config: {e}")
        return {}


def get_current_model(config_data):
    """Extract current model from config data."""
    if not config_data:
        return ''

    # Try TOML format first
    model = config_data.get('model', {})
    if isinstance(model, dict):
        return model.get('default', '')

    # Try YAML format (nested structure)
    providers = config_data.get('providers', {})
    if providers:
        # Get first provider's default model
        for provider_name, provider_config in providers.items():
            if isinstance(provider_config, dict):
                default_model = provider_config.get('default_model', '')
                if default_model:
                    return default_model

    return ''


def update_config_yaml(file_path, new_model):
    """Update the model in config.yaml."""
    try:
        with open(file_path, 'r') as f:
            config = yaml.safe_load(f)

        if not config:
            config = {}

        # Update model.default or providers[*].default_model
        if 'model' in config and isinstance(config['model'], dict):
            config['model']['default'] = new_model
        elif 'providers' in config and isinstance(config['providers'], dict):
            for provider_name, provider_config in config['providers'].items():
                if isinstance(provider_config, dict) and 'default_model' in provider_config:
                    provider_config['default_model'] = new_model
                    break

        with open(file_path, 'w') as f:
            yaml.dump(config, f, default_flow_style=False, allow_unicode=True)

        return {'success': True, 'message': f'Updated config to model: {new_model}'}
    except Exception as e:
        return {'success': False, 'error': str(e)}


def find_provider(model_name, config_data):
    """Find the provider for a given model.

    Priority:
    1. Top-level model.default (the actual active model)
    2. Provider's models dict
    3. Provider's default_model
    4. Fallback: match by model name patterns (e.g., agnes-* → agnes provider)
    """
    if not model_name or not config_data:
        return None

    providers = config_data.get('providers', {})

    # 1. Check top-level model.default — this is the actual active model
    top_model = config_data.get('model', {})
    if isinstance(top_model, dict):
        top_default = top_model.get('default', '')
        if top_default == model_name:
            for pname, pconfig in providers.items():
                if isinstance(pconfig, dict) and pconfig.get('api'):
                    return {
                        'name': pname,
                        'api': pconfig['api'],
                        'models': [model_name]
                    }
            return None

    # 2. Check provider models dict
    if providers:
        for provider_name, provider_config in providers.items():
            if isinstance(provider_config, dict):
                models = provider_config.get('models', {})
                if isinstance(models, dict) and model_name in models:
                    return {
                        'name': provider_name,
                        'api': provider_config.get('api', ''),
                        'models': list(models.keys())
                    }

                # 3. Check provider default_model
                if provider_config.get('default_model') == model_name:
                    return {
                        'name': provider_name,
                        'api': provider_config.get('api', ''),
                        'models': [model_name]
                    }

    # 4. Fallback: match by model name patterns
    if providers:
        for provider_name, provider_config in providers.items():
            if isinstance(provider_config, dict) and provider_config.get('api'):
                # agnes-* → agnes provider
                if model_name.lower().startswith(provider_name.lower()):
                    return {
                        'name': provider_name,
                        'api': provider_config['api'],
                        'models': [model_name]
                    }

    return None


def add_history(action, from_model, to_model, mode, success):
    """Add entry to switch history."""
    entry = {
        'timestamp': int(time.time() * 1000),
        'action': action,
        'from': from_model,
        'to': to_model,
        'mode': mode,
        'success': success
    }
    switch_history.append(entry)
    if len(switch_history) > MAX_HISTORY:
        switch_history.pop(0)
    return entry


# API Routes

@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint."""
    config_path = find_config_yaml()
    config_data = parse_config_yaml(config_path) if os.path.exists(config_path) else {}
    current_model = get_current_model(config_data)

    return jsonify({
        'status': 'healthy',
        'uptime': time.time(),
        'current_model': current_model,
        'config_path': config_path
    })


@app.route('/v1/models', methods=['GET'])
def list_models():
    """List available models."""
    config_path = find_config_yaml()
    config_data = parse_config_yaml(config_path) if os.path.exists(config_path) else {}

    models = []
    providers = config_data.get('providers', {})

    if providers:
        for provider_name, provider_config in providers.items():
            if isinstance(provider_config, dict):
                provider_models = provider_config.get('models', {})
                if isinstance(provider_models, dict):
                    for model_name in provider_models.keys():
                        models.append({
                            'id': model_name,
                            'object': 'model',
                            'created': int(time.time()),
                            'owned_by': provider_name
                        })

    return jsonify({
        'object': 'list',
        'data': models
    })


@app.route('/api/config', methods=['GET'])
def get_config():
    """Get current configuration."""
    config_path = find_config_yaml()
    config_data = parse_config_yaml(config_path) if os.path.exists(config_path) else {}
    current_model = get_current_model(config_data)

    return jsonify({
        'model': current_model,
        'path': config_path,
        'raw': config_data
    })


@app.route('/api/routing-mode', methods=['GET'])
def get_routing_mode():
    """Get current routing mode."""
    load_routing_mode()
    return jsonify({'mode': routing_mode})


@app.route('/api/set-routing-mode', methods=['POST'])
def set_routing_mode():
    """Set routing mode."""
    data = request.get_json()
    mode = data.get('mode') if data else None

    if mode not in ['codex', 'config', 'both']:
        return jsonify({'success': False, 'error': 'Invalid mode', 'code': 'INVALID_MODE'}), 400

    if update_routing_mode(mode):
        return jsonify({'success': True, 'mode': mode})
    else:
        return jsonify({'success': False, 'error': 'Failed to update routing mode'}), 500


@app.route('/api/providers/status', methods=['GET'])
def get_providers_status():
    """Get provider status."""
    config_path = find_config_yaml()
    config_data = parse_config_yaml(config_path) if os.path.exists(config_path) else {}

    providers = []
    provider_configs = config_data.get('providers', {})

    if provider_configs:
        for name, config in provider_configs.items():
            if isinstance(config, dict):
                api_url = config.get('api', '')
                # Check if API key is configured
                status = 'online' if api_url else 'offline'
                providers.append({
                    'name': name,
                    'baseUrl': api_url,
                    'status': status,
                    'models': list(config.get('models', {}).keys()) if isinstance(config.get('models'), dict) else []
                })

    return jsonify({'providers': providers})


KNOWN_BALANCE_DOMAINS = {
    'api.openai.com': ('GET', '/dashboard/billing/credit_grants'),
    'api.anthropic.com': None,  # No public balance API
    'api.deepseek.com': ('GET', '/user/info'),
    'api.moonshot.cn': ('GET', '/user/info'),
}


@app.route('/api/balances', methods=['GET'])
def get_balances():
    """Get account balances (simplified - actual implementation depends on provider APIs)."""
    balances = {}
    config_path = find_config_yaml()
    config_data = parse_config_yaml(config_path) if os.path.exists(config_path) else {}

    providers = config_data.get('providers', {})
    for name, config in providers.items():
        if not isinstance(config, dict):
            balances[name] = '不可查询'
            continue
        api_url = config.get('api', '')
        if not api_url:
            balances[name] = '不可查询'
            continue
        # Extract hostname and check against known providers
        from urllib.parse import urlparse
        try:
            parsed = urlparse(api_url)
            host = parsed.hostname or ''
        except Exception:
            host = ''
        if host in KNOWN_BALANCE_DOMAINS:
            endpoint = KNOWN_BALANCE_DOMAINS[host]
            if endpoint is None:
                balances[name] = '无公开余额API'
            else:
                method, path = endpoint
                try:
                    url = f"{api_url.rstrip('/')}{path}"
                    resp = requests.get(url, timeout=5)
                    if resp.ok:
                        data = resp.json()
                        balances[name] = data.get('balance', data.get('amount', data.get('totalAmount', '未知')))
                    else:
                        balances[name] = f'查询失败 (HTTP {resp.status_code})'
                except Exception as e:
                    balances[name] = f'查询失败: {str(e)}'
        else:
            balances[name] = '不可查询'

    return jsonify({'balances': balances})


@app.route('/api/history', methods=['GET'])
def get_history():
    """Get switch history."""
    return jsonify({
        'history': switch_history[:20][::-1]  # Last 20 entries, reversed
    })


@app.route('/api/switch-model', methods=['POST'])
def switch_model():
    """Switch to a different model."""
    data = request.get_json()
    if not data or 'model' not in data:
        return jsonify({'success': False, 'error': 'Missing model parameter', 'code': 'MISSING_MODEL'}), 400

    new_model = data['model']
    config_path = find_config_yaml()
    config_data = parse_config_yaml(config_path) if os.path.exists(config_path) else {}

    old_model = get_current_model(config_data)

    # Validate model exists
    provider = find_provider(new_model, config_data)
    if not provider:
        return jsonify({'success': False, 'error': f'Unsupported model: {new_model}', 'code': 'UNSUPPORTED_MODEL'}), 400

    # Update config
    result = update_config_yaml(config_path, new_model)
    if not result['success']:
        return jsonify({'success': False, 'error': result['error'], 'code': 'CONFIG_UPDATE_FAILED'}), 500

    # Auto-switch to config routing mode
    update_routing_mode('config')

    # Record history
    add_history('切换模型', old_model, new_model, 'config', True)

    logger.info(f"[ADMIN] Model switched: {old_model} → {new_model}")

    return jsonify({
        'success': True,
        'message': f'Successfully switched to {new_model}, routing mode set to config',
        'oldModel': old_model,
        'newModel': new_model,
        'needRestart': True
    })


@app.route('/api/test-connection', methods=['POST'])
def test_connection():
    """Test connection to a model's provider."""
    data = request.get_json()
    if not data or 'model' not in data:
        return jsonify({'success': False, 'error': 'Missing model parameter', 'code': 'MISSING_MODEL'}), 400

    model = data['model']
    config_path = find_config_yaml()
    config_data = parse_config_yaml(config_path) if os.path.exists(config_path) else {}

    provider = find_provider(model, config_data)
    if not provider:
        return jsonify({'success': False, 'error': f'Unsupported model: {model}'}), 400

    try:
        # Try to fetch models from provider API
        api_url = provider['api']
        if api_url:
            response = requests.get(f"{api_url}/models", timeout=10)
            if response.ok:
                return jsonify({'success': True, 'message': 'Connection successful'})
            else:
                return jsonify({'success': False, 'error': f'HTTP {response.status_code}', 'code': 'UPSTREAM_HTTP_ERROR'}), 200
        else:
            return jsonify({'success': False, 'error': 'No API URL configured', 'code': 'NO_API_URL'}), 200

    except Exception as e:
        return jsonify({'success': False, 'error': str(e), 'code': 'CONNECTION_FAILED'})


@app.route('/api/clear-history', methods=['POST'])
def clear_history():
    """Clear switch history."""
    global switch_history
    switch_history = []
    return jsonify({'success': True})


# Main entry point
if __name__ == '__main__':
    logger.info(f"Hermes Multi-Model Proxy starting on port {PORT}")
    logger.info(f"Config path: {find_config_yaml()}")

    # Load initial routing mode
    load_routing_mode()

    app.run(host='127.0.0.1', port=PORT, debug=False)

File: codex-proxy/proxy.js
const express = require('express');
require('dotenv').config();
const fs = require('fs');
const path = require('path');

const app = express();
app.use(express.json({ limit: '50mb' }));
app.use(express.urlencoded({ extended: true, limit: '50mb' }));

console.log = (...args) => {
  const ts = new Date().toISOString();
  process.stdout.write(`[${ts}] ${args.join(' ')}\n`);
};

const PORT = parseInt(process.env.PORT) || 18790;
const HOME = process.env.HOME || path.join(require('os').homedir());

// ===== Config file detection =====
const ROUTING_MODE_FILE = path.join(HOME, '.codex-proxy', 'routing-mode.json');

let CONFIG_TOML = findConfigToml();
let lastConfigModel = '';
let routingMode = 'codex'; // 'codex' | 'config' | 'both'

function findConfigToml() {
  const home = HOME;
  const candidates = [
    path.join(home, '.hermes', 'config.toml'),
    path.join(home, 'Library', 'Containers', 'app.nousresearch.hermes', 'Data', '.hermes', 'config.yaml'),
    path.join(home, '.codex', 'config.toml'),
    path.join(home, 'Library', 'Containers', 'com.openai.codex', 'Data', '.codex', 'config.toml'),
    path.join(home, 'Library', 'Application Support', 'Codex', 'config.toml'),
    path.join(home, '.codex-plus-plus', 'config.toml')
  ];

  for (const candidate of candidates) {
    try {
      if (fs.existsSync(candidate)) {
        console.log(`[CONFIG] 检测到配置文件: ${candidate}`);
        return candidate;
      }
    } catch (e) {
      // ignore
    }
  }
  return candidates[0];
}

function parseConfigToml(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    const modelMatch = content.match(/^model\s*=\s*"([^"]+)"/m);
    const modelProviderMatch = content.match(/^model_provider\s*=\s*"([^"]+)"/m);
    const wireApiMatch = content.match(/^wire_api\s*=\s*"([^"]+)"/m);
    const yamlModelMatch = content.match(/(?:^|\n)\s*default:\s*(\S+)/m);
    const yamlProviderMatch = content.match(/(?:^|\n)\s*provider:\s*(\S+)/m);

    let model = '';
    let model_provider = '';
    let wire_api = '';

    if (filePath.endsWith('.toml') || modelMatch) {
      model = modelMatch ? modelMatch[1] : '';
      model_provider = modelProviderMatch ? modelProviderMatch[1] : '';
      wire_api = wireApiMatch ? wireApiMatch[1] : '';
    } else if (filePath.endsWith('.yaml') || filePath.endsWith('.yml') || yamlModelMatch) {
      model = yamlModelMatch ? yamlModelMatch[1] : '';
      model_provider = yamlProviderMatch ? yamlProviderMatch[1] : '';
      wire_api = '';
    }

    return { model, model_provider, wire_api, raw: content };
  } catch (e) {
    return { model: '', model_provider: '', wire_api: '', raw: '', error: e.message };
  }
}

function updateConfigToml(filePath, newModel) {
  try {
    let content = fs.readFileSync(filePath, 'utf-8');
    if (filePath.endsWith('.toml')) {
      content = content.replace(/^model\s*=\s*".+"$/m, `model = "${newModel}"`);
    } else if (filePath.endsWith('.yaml') || filePath.endsWith('.yml')) {
      content = content.replace(/(default:\s*)(\S+)/m, `$1${newModel}`);
      if (!content.includes(`default: ${newModel}`)) {
        content = content.replace(/(model:\s*\n\s*default:\s*)(\S+)/m, `$1${newModel}`);
      }
    } else {
      content = content.replace(/^model\s*=\s*".+"$/m, `model = "${newModel}"`);
    }
    fs.writeFileSync(filePath, content, 'utf-8');
    return { success: true, message: `已更新配置文件，model 设置为 ${newModel}` };
  } catch (e) {
    return { success: false, error: e.message };
  }
}

function getConfigModel() {
  if (!CONFIG_TOML) return '';
  try {
    const cfg = parseConfigToml(CONFIG_TOML);
    const newModel = cfg.model || '';
    if (newModel && newModel !== lastConfigModel) {
      if (lastConfigModel) {
        console.log(`[CONFIG WATCH] ${CONFIG_TOML} 变更: "${lastConfigModel}" → "${newModel}"`);
      }
      lastConfigModel = newModel;
      return newModel;
    }
  } catch (e) {
    console.warn(`[CONFIG WATCH] 读取 ${CONFIG_TOML} 失败: ${e.message}`);
  }
  return lastConfigModel;
}

function loadRoutingMode() {
  try {
    const data = fs.readFileSync(ROUTING_MODE_FILE, 'utf-8');
    const parsed = JSON.parse(data);
    if (parsed.mode && ['codex', 'config', 'both'].includes(parsed.mode)) {
      routingMode = parsed.mode;
    }
  } catch {
    routingMode = 'codex';
  }
}

function updateRoutingMode(mode) {
  try {
    const dir = path.dirname(ROUTING_MODE_FILE);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    const data = JSON.stringify({ mode }, null, 2);
    const tmpFile = ROUTING_MODE_FILE + '.tmp';
    fs.writeFileSync(tmpFile, data, 'utf-8');
    fs.renameSync(tmpFile, ROUTING_MODE_FILE);
    console.log(`[ROUTING MODE] 路由模式已更新为: ${mode}`);
    return true;
  } catch (e) {
    // Clean up temp file on failure
    try { fs.unlinkSync(ROUTING_MODE_FILE + '.tmp'); } catch {}
    console.error(`[ROUTING MODE] 更新路由模式失败: ${e.message}`);
    return false;
  }
}

// ===== Upstream models =====
const UPSTREAM_MODELS = [
  {
    name: 'DeepSeek',
    baseUrl: 'https://api.deepseek.com/v1',
    apiKey: process.env.DEEPSEEK_API_KEY,
    availableModels: ['deepseek-v4-pro', 'deepseek-v4-flash'],
  },
  {
    name: 'Kimi',
    baseUrl: 'https://api.moonshot.cn/v1',
    apiKey: process.env.MOONSHOT_API_KEY,
    availableModels: ['moonshot-v1-8k', 'moonshot-v1-32k', 'moonshot-v1-128k', 'kimi-k2.5', 'kimi-k2.6', 'kimi-k2.7-code', 'kimi-k2.7-code-highspeed', 'moonshot-v1-128k-vision-preview', 'moonshot-v1-32k-vision-preview', 'moonshot-v1-8k-vision-preview', 'moonshot-v1-auto']
  },
  {
    name: 'Agnes',
    baseUrl: 'https://apihub.agnes-ai.com/v1',
    apiKey: process.env.AGNES_API_KEY,
    availableModels: ['agnes-2.0-flash', 'agnes-1.5-flash', 'agnes-image-2.0-flash', 'agnes-image-2.1-flash', 'agnes-video-v2.0']
  }
];

function findProvider(modelName) {
  if (!modelName) return null;
  let provider = UPSTREAM_MODELS.find(p => p.availableModels.includes(modelName));
  if (provider) return provider;
  const lower = modelName.toLowerCase();
  if (lower.includes('deepseek')) return UPSTREAM_MODELS[0];
  if (lower.includes('kimi') || lower.includes('moonshot')) return UPSTREAM_MODELS[1];
  if (lower.includes('agnes')) return UPSTREAM_MODELS[2];
  return null;
}

// ===== API Routes =====

// Models list
app.get('/v1/models', (req, res) => {
  const models = UPSTREAM_MODELS.flatMap(p =>
    p.availableModels.map(name => ({
      id: name, object: 'model', created: Math.floor(Date.now() / 1000), owned_by: p.name
    }))
  );
  res.json({ object: 'list', data: models });
});

// Health check
app.get('/health', (req, res) => {
  res.json({
    status: 'healthy',
    uptime: process.uptime(),
    models: UPSTREAM_MODELS.flatMap(p => p.availableModels)
  });
});

// Admin API routes
app.get('/api/config', (req, res) => {
  const configTomlPath = findConfigToml();
  const config = parseConfigToml(configTomlPath);
  config.path = configTomlPath;
  res.json(config);
});

app.get('/api/routing-mode', (req, res) => {
  loadRoutingMode();
  res.json({ mode: routingMode });
});

app.post('/api/set-routing-mode', (req, res) => {
  const { mode } = req.body;
  if (!['codex', 'config', 'both'].includes(mode)) {
    return res.status(400).json({ success: false, error: 'Invalid mode', code: 'INVALID_MODE' });
  }
  try {
    if (!updateRoutingMode(mode)) {
      return res.status(500).json({ success: false, error: 'Failed to write routing mode file', code: 'WRITE_FAILED' });
    }
    routingMode = mode;
    console.log(`[ADMIN] 路由模式设置为: ${mode}`);
    res.json({ success: true, mode });
  } catch (e) {
    console.error(`[ADMIN] 设置路由模式失败: ${e.message}`);
    res.status(500).json({ success: false, error: e.message, code: 'INTERNAL_ERROR' });
  }
});

app.get('/api/providers/status', (req, res) => {
  const providers = UPSTREAM_MODELS.map(p => ({
    name: p.name,
    baseUrl: p.baseUrl,
    status: p.apiKey ? 'online' : 'offline',
    models: p.availableModels
  }));
  res.json({ providers });
});

app.get('/api/balances', async (req, res) => {
  const balances = {};
  for (const provider of UPSTREAM_MODELS) {
    if (!provider.apiKey) { balances[provider.name] = '未配置'; continue; }
    try {
      let url;
      if (provider.name === 'DeepSeek') url = `${provider.baseUrl}/user/info`;
      else if (provider.name === 'Kimi') url = `${provider.baseUrl}/user/info`;
      else if (provider.name === 'Agnes') url = `${provider.baseUrl}/user/balance`;
      else { balances[provider.name] = '不支持'; continue; }

      const response = await fetch(url, {
        headers: { 'Authorization': `Bearer ${provider.apiKey}` }
      });
      if (response.ok) {
        const data = await response.json();
        balances[provider.name] = data.balance || data.amount || '未知';
      } else {
        balances[provider.name] = '查询失败';
      }
    } catch (e) {
      balances[provider.name] = '错误';
    }
  }
  res.json({ balances });
});

// Switch history (in-memory)
const switchHistory = [];
const MAX_HISTORY = 50;

app.get('/api/history', (req, res) => {
  res.json({ history: switchHistory.slice(0, 20).reverse() });
});

app.post('/api/switch-model', (req, res) => {
  const { model } = req.body;
  if (!model) {
    return res.status(400).json({ success: false, error: '缺少 model 参数', code: 'MISSING_MODEL' });
  }
  const provider = findProvider(model);
  if (!provider) {
    return res.status(400).json({ success: false, error: `不支持的模型: ${model}`, code: 'UNSUPPORTED_MODEL' });
  }
  try {
    const configTomlPath = findConfigToml();
    const currentConfig = parseConfigToml(configTomlPath);
    const oldModel = currentConfig.model;

    const updateResult = updateConfigToml(configTomlPath, model);
    if (!updateResult.success) throw new Error(updateResult.error);

    updateRoutingMode('config');
    routingMode = 'config';

    console.log(`[ADMIN] 模型切换: ${oldModel} → ${model}`);
    addHistory('切换模型', oldModel, model, 'config', true);

    res.json({
      success: true,
      message: `已成功切换到 ${model}，路由模式已设置为 config`,
      oldModel,
      newModel: model,
      needRestart: true
    });
  } catch (e) {
    console.error(`[ADMIN] 切换模型失败: ${e.message}`);
    addHistory('切换模型', '', model, '', false);
    res.status(500).json({ success: false, error: e.message, code: 'SWITCH_FAILED' });
  }
});

app.post('/api/test-connection', async (req, res) => {
  const { model } = req.body;
  if (!model) {
    return res.status(400).json({ success: false, error: '缺少 model 参数', code: 'MISSING_MODEL' });
  }
  const provider = findProvider(model);
  if (!provider) {
    return res.status(400).json({ success: false, error: `不支持的模型: ${model}`, code: 'UNSUPPORTED_MODEL' });
  }
  try {
    const response = await fetch(`${provider.baseUrl}/models`, {
      headers: { 'Authorization': `Bearer ${provider.apiKey}` }
    });
    if (response.ok) {
      res.json({ success: true, message: '连接成功' });
    } else {
      res.json({ success: false, error: `HTTP ${response.status}`, code: 'UPSTREAM_HTTP_ERROR' });
    }
  } catch (e) {
    res.json({ success: false, error: e.message, code: 'CONNECTION_FAILED' });
  }
});

app.post('/api/clear-history', (req, res) => {
  switchHistory.length = 0;
  res.json({ success: true });
});

function addHistory(action, from, to, mode, success) {
  switchHistory.push({ timestamp: Date.now(), action, from, to, mode, success });
  if (switchHistory.length > MAX_HISTORY) switchHistory.shift();
}

// Chat Completions passthrough
app.post('/v1/chat/completions', async (req, res) => {
  try {
    const { model } = req.body;
    const provider = findProvider(model);
    if (!provider) {
      return res.status(400).json({ success: false, error: `Unsupported model: ${model}`, code: 'UNSUPPORTED_MODEL' });
    }

    const upstreamUrl = `${provider.baseUrl}/chat/completions`;
    const response = await fetch(upstreamUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${provider.apiKey}`
      },
      body: JSON.stringify(req.body),
      signal: AbortSignal.timeout(120000),
    });

    if (!response.ok) {
      const err = await response.text();
      console.error(`Chat upstream error: ${err}`);
      return res.status(response.status).json({ success: false, error: err, code: 'UPSTREAM_ERROR' });
    }

    if (req.body.stream) {
      res.setHeader('Content-Type', 'text/event-stream');
      res.setHeader('Cache-Control', 'no-cache');
      res.setHeader('Connection', 'keep-alive');
      try {
        await response.body.pipeTo(res);
      } catch (err) {
        console.error(`[Stream] Pipe error: ${err.message}`);
        if (!res.headersSent) {
          res.status(504).json({ success: false, error: 'Stream timeout or upstream disconnected', code: 'STREAM_TIMEOUT' });
        }
      }
    } else {
      const data = await response.json();
      res.json(data);
    }
  } catch (e) {
    console.error(`Chat error: ${e.message}`);
    res.status(500).json({ success: false, error: e.message, code: 'CHAT_ERROR' });
  }
});

// Listen
app.listen(PORT, () => {
  console.log(`Codex Multi-Model Proxy started on port ${PORT}`);
  console.log(`Config: ${CONFIG_TOML}`);
  console.log(`Models: ${UPSTREAM_MODELS.flatMap(p => p.availableModels).join(', ')}`);
  loadRoutingMode();
});
